[WIP] ChartEditor: use react-chart-editor

n-riesco opened the pull request:
@jackparmer This is just to show you how Falcon would look like with `react-chart-editor`.

See that I've made the Table view the default.

I'm also thinking of making the tree view as big as the code editor. So that the chart editor can use the whole width.

@nicolaskruchten I'm looking forward to the new API (it'd simplify this PR a lot).

![peek 2018-03-13 17-45](https://user-images.githubusercontent.com/6199391/37359812-782f9a1c-26e6-11e8-9b3f-e48275e62271.gif)

jackparmer commented:
Nice. It looks great. When you switch to the chart tab, should we
auto-minimize the table schema? `react-chart-editor` needs a good amount of
horizontal space and users might not now that they can collapse the table
schema sidebar.


On Tue, Mar 13, 2018 at 1:51 PM, Nicolas Riesco <notifications@github.com>
wrote:

> @jackparmer <https://github.com/jackparmer> This is just to show you how
> Falcon would look like with react-chart-editor.
>
> See that I've made the Table view the default.
>
> I'm also thinking of making the tree view as big as the code editor. So
> that the chart editor can use the whole width.
>
> @nicolaskruchten <https://github.com/nicolaskruchten> I'm looking forward
> to the new API (it'd simplify this PR a lot).
>
> [image: peek 2018-03-13 17-45]
> <https://user-images.githubusercontent.com/6199391/37359812-782f9a1c-26e6-11e8-9b3f-e48275e62271.gif>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/plotly/falcon-sql-client/pull/405
> Commit Summary
>
>    - ChartEditor: WIP
>
> File Changes
>
>    - *M* app/components/Settings/Preview/Preview.react.js
>    <https://github.com/plotly/falcon-sql-client/pull/405/files#diff-0>
>    (24)
>    - *A* app/components/Settings/Preview/chart-editor.jsx
>    <https://github.com/plotly/falcon-sql-client/pull/405/files#diff-1>
>    (134)
>    - *M* package.json
>    <https://github.com/plotly/falcon-sql-client/pull/405/files#diff-2>
>    (6)
>    - *M* webpack.config.base.js
>    <https://github.com/plotly/falcon-sql-client/pull/405/files#diff-3>
>    (7)
>    - *M* yarn.lock
>    <https://github.com/plotly/falcon-sql-client/pull/405/files#diff-4>
>    (1015)
>
> Patch Links:
>
>    - https://github.com/plotly/falcon-sql-client/pull/405.patch
>    - https://github.com/plotly/falcon-sql-client/pull/405.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/plotly/falcon-sql-client/pull/405>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABx4apFPzoyZIzzO4peWgRudX5EiAO8Bks5teAcqgaJpZM4SpKgR>
> .
>

nicolaskruchten commented:
Awesome! New release coming soon with the new API /cc @VeraZab

n-riesco commented:
@jackparmer I thought about autominimise, but I was concerned some users might find difficult to discover that they can make the schema view visible again. That's why I thought making the schema view as big as the code editor would be the solution.

jackparmer commented:
> That's why I thought making the schema view as big as the code editor
would be the solution.

Yeah, that sounds good to me.

On Tue, Mar 13, 2018 at 2:10 PM, Nicolas Riesco <notifications@github.com>
wrote:

> @jackparmer <https://github.com/jackparmer> I thought about autominimise,
> but I was concerned some users might find difficult to discover that they
> can make the schema view visible again. That's why I thought making the
> schema view as big as the code editor would be the solution.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/plotly/falcon-sql-client/pull/405#issuecomment-372764790>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABx4ajQCXkpwdfd467hxHfFGJ06oBpm_ks5teAuwgaJpZM4SpKgR>
> .
>

VeraZab commented:
@n-riesco just made a new release: https://www.npmjs.com/package/react-chart-editor
cc @nicolaskruchten

n-riesco commented:
@jackparmer I've updated the PR. In this end I've gone for the option of hiding both the code editor and the table schemas view (it didn't make sense to hide the code editor while the table schemas view was still visible)

@nicolaskruchten @VeraZab As you can see in the screencast, I've got a problem with the initial plot size. Is there a way to set this size with the new API?

![peek 2018-03-15 20-05](https://user-images.githubusercontent.com/6199391/37488516-d9fc6548-288c-11e8-8b19-39fc407c64df.gif)

nicolaskruchten commented:
`react-plotly.js@2.1.0` and `react-chart-editor@0.13.0` should resolve the sizing issue!

n-riesco commented:
@nicolaskruchten One thing, that is rather inconvenient with this workaround, is that I have to hard-code the height as `style={{height: 400}}`. I tried `style={{minHeight: 400, height: '100%'}}`, but  it wouldn't work (the height of `<PlotlyEditor>` would be less than `400px`).

![peek 2018-03-16 21-12](https://user-images.githubusercontent.com/6199391/37545043-1fe00f92-295f-11e8-9319-095ba88d454b.gif)

nicolaskruchten commented:
What's the easiest way for me to get this running locally so I can play with the CSS? The height/width thing is a big pain point with this project in general :(

jackparmer commented:
@nicolaskruchten This works for me: https://github.com/plotly/falcon-sql-client/blob/master/CONTRIBUTING.md

n-riesco commented:
@nicolaskruchten It turns out the issue isn't with the CSS. In fact,  `style={{minHeight: 400, height: '100%'}}` works if I'm careful to mount `<PlotEditor>` only when the parent node is visible (otherwise `this.PlotComponent` throws errors when `window.getComputedStyle(gd).display === 'none`).

I have something working, but I'll clean it and push it into the PR tomorrow.

n-riesco commented:
@nicolaskruchten In principle https://github.com/plotly/falcon-sql-client/pull/405/commits/85afe3b32e402f794a472fda92d7b070398d5cf4 could be implemented inside `react-chart-editor`, but this solution isn't very efficient.

I think the place to deal with this issue is inside `plotly.js` (I don't know what the implications are, but ideally we should be able to tell `plotly.js` to ignore `display: none;` and draw the chart as if it was visible).

nicolaskruchten commented:
Seconded, this is a problem in the editor in general but I thought I'd more or less fixed it by [firing `window.resize` after drawing](https://github.com/plotly/react-plotly.js/commit/ca8ece2e597cbe2e2175dec1eeb722ef5e39afa0#diff-d4165b1e3527a04dd7e9d370e18c2a1aR164). Clearly that's not late enough in the React loop to catch all cases. It's happening in Dash-land as well: https://plotly.slack.com/archives/C90K5TYP3/p1521599318000139

n-riesco commented:
@nicolaskruchten thanks for reviewing https://github.com/n-riesco/plotly-database-connector/blob/04dd2463ccddf834b19ef4111cffa989bf7f7fa6/app/components/Settings/Preview/chart-editor.jsx . I've renamed `gd` back to `plotlyJSON` to make clear this isn't the DOM element.

n-riesco commented:
@shannonlal I'm done with this PR. Would you review it, please?

shannonlal commented:
@n-riesco Looking at this now

shannonlal commented on app/components/Settings/Preview/chart-editor.jsx:
```diff
@@ -0,0 +1,137 @@
+        });
+
+        // Cap plots to 100k rows
+        const length = Math.min(rows.length, 100000);
```
This should probably be defined as a const at top.  const MAX_ROWS=10000;

shannonlal commented on app/components/Settings/Preview/chart-editor.jsx:
```diff
@@ -0,0 +1,137 @@
+            <div
+                ref={'container'}
+                style={{
+                    minHeight: 400,
```
Just a suggestion.  Maybe a constant or move to a separate file.  I see the min height of 400 defined in a bunch of files.  Maybe we should centralize to a single file?  your thoughts?

n-riesco replied:
I've re-enabled [css-loader](https://github.com/webpack-contrib/css-loader), so that we can import again `.css` files into a component. But I didn't create a `chart-editor.css`, because I feel we need tidy up the CSS in Falcon first:
- perhaps define a `.css` file with a common theme
- follow [Plotly's style guide](https://brand.plot.ly/)
- Jack mentioned  [tachyons](http://tachyons.io/) and I'm keen to give it a go.

shannonlal commented on app/components/Settings/Settings.react.js:
```diff
@@ -6,14 +6,12 @@ import ReactToolTip from 'react-tooltip';
 import DialectSelector from './DialectSelector/DialectSelector.react';
 import ConnectButton from './ConnectButton/ConnectButton.react';
 import Preview from './Preview/Preview.react';
-import TableTree from './Preview/TableTree.react.js';
```
Is this file no longer in use?  I was working on some jest tests for table tree.  If it is not needed I can move on to some other tests

n-riesco replied:
yes, it's still in use (tests are always welcome :smile:). I moved the import to `Preview.react.js`. Now `Preview.react.js` contains all the components displayed under the `QUERY` tab.

shannonlal commented on test/app/components/Settings/Preview/chart-editor.test.jsx:
```diff
@@ -0,0 +1,87 @@
+        columnNames = ['x', 'y'];
+
+        rows = [];
+        for (let i = 0; i < 100001; i++) {
```
Should this be (MAX_LENGTH + 1)

n-riesco commented:
@shannonlal I've addressed all your comments except that about CSS (I'd rather address it after we devise a plan to tidy up the CSS in Falcon; see what I replied [here](https://github.com/plotly/falcon-sql-client/pull/405#discussion_r176913259)).

zhaodagang commented:
Can this App connect clickhouse database ?

2018-05-09 

zdg8882000 



发件人：Nicolas Riesco <notifications@github.com>
发送时间：2018-05-08 23:43
主题：Re: [plotly/falcon-sql-client] ChartEditor: use react-chart-editor (#405)
收件人："plotly/falcon-sql-client"<falcon-sql-client@noreply.github.com>
抄送："Mr Zhao"<zdg8882000@163.com>,"Mention"<mention@noreply.github.com>

@zhaodagang Falcon and react-chart-editor can be used without login.
react-chart-editor is based on the open-source library plotly.js.
If you create an account on https://plot.ly/ and login with Falcon, then you'll be able to upload your data and charts onto https://plot.ly/ .
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.
